baseplate/frameworks/pyramid/__init__.py

```diff
@@ -166,29 +171,83 @@ def baseplate_tween(request: Request) -> Response:
                 "http_endpoint": http_endpoint,
                 "http_success": http_success,
             }
-            ACTIVE_REQUESTS.labels(http_method=http_method, http_endpoint=http_endpoint).dec()
-            REQUEST_LATENCY.labels(**histogram_labels).observe(time.perf_counter() - time_started)
-            REQUESTS_TOTAL.labels(
-                **{
-                    **histogram_labels,
-                    "http_response_code": http_response_code,
-                }
-            ).inc()
-            REQUEST_SIZE.labels(**histogram_labels).observe(request.content_length or 0)
-            if response:
-                try:
-                    if response.content_length is not None:
-                        RESPONSE_SIZE.labels(**histogram_labels).observe(response.content_length)
-                except Exception:
-                    pass
+
+            if getattr(request, "prom_metrics_enabled", False):
+                ACTIVE_REQUESTS.labels(http_method=http_method, http_endpoint=http_endpoint).dec()
+                REQUEST_LATENCY.labels(**histogram_labels).observe(
+                    time.perf_counter() - time_started
+                )
+                REQUESTS_TOTAL.labels(
+                    **{
+                        **histogram_labels,
+                        "http_response_code": http_response_code,
+                    }
+                ).inc()
+                REQUEST_SIZE.labels(**histogram_labels).observe(request.content_length or 0)
+                if response:
+                    try:
+                        if response.content_length is not None:
+                            RESPONSE_SIZE.labels(**histogram_labels).observe(
+                                response.content_length
+                            )
+                    except Exception:
+                        pass
 
             # avoid a reference cycle
+            request.prom_metrics_enabled = False
             request.start_server_span = None
         return response
 
     return baseplate_tween
 
 
+def _manually_close_request_metrics(request: Request) -> None:
```

Comment:
This looks like a lot of code duplication. What prevents us from calling this in the tween above?

Reply:
it is partially, but this has specific handling around not having a response to process things like sizes and response codes from

Reply:
Consolidated the logic